tests/unit/s2n_renegotiate_test.c

```diff
@@ -0,0 +1,402 @@
+            EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&in, 0));
+            EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&out, 0));
+            EXPECT_SUCCESS(s2n_connection_set_io_stuffers(&in, &out, client_conn));
+            EXPECT_SUCCESS(s2n_connection_set_io_stuffers(&out, &in, server_conn));
+
+            EXPECT_SUCCESS(s2n_connection_prefer_low_latency(client_conn));
+            EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+
+            EXPECT_EQUAL(s2n_send(client_conn, app_data, sizeof(app_data), &blocked), sizeof(app_data));
+            size_t original_out_size = s2n_stuffer_data_available(&out);
+            EXPECT_SUCCESS(s2n_stuffer_wipe(&out));
+
+            EXPECT_SUCCESS(s2n_renegotiate_wipe(client_conn));
+
+            EXPECT_EQUAL(s2n_send(client_conn, app_data, sizeof(app_data), &blocked), sizeof(app_data));
+            size_t wiped_out_size = s2n_stuffer_data_available(&out);
+            EXPECT_EQUAL(original_out_size, wiped_out_size);
+        }
+
+        /* Forced very small fragment size unaffected by wipe */
+        {
+            DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER), s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(server_conn);
+            EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+
+            DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT), s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(client_conn);
+            EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+
+            DEFER_CLEANUP(struct s2n_stuffer in = { 0 }, s2n_stuffer_free);
+            DEFER_CLEANUP(struct s2n_stuffer out = { 0 }, s2n_stuffer_free);
+            EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&in, 0));
+            EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&out, 0));
+            EXPECT_SUCCESS(s2n_connection_set_io_stuffers(&in, &out, client_conn));
+            EXPECT_SUCCESS(s2n_connection_set_io_stuffers(&out, &in, server_conn));
+
+            EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+
+            size_t small_frag_len = 1;
+            client_conn->max_outgoing_fragment_length = small_frag_len;
+
+            EXPECT_EQUAL(s2n_send(client_conn, app_data, sizeof(app_data), &blocked), sizeof(app_data));
+            size_t original_out_size = s2n_stuffer_data_available(&out);
+            EXPECT_SUCCESS(s2n_stuffer_wipe(&out));
+
+            EXPECT_SUCCESS(s2n_renegotiate_wipe(client_conn));
+            EXPECT_EQUAL(client_conn->max_outgoing_fragment_length, small_frag_len);
+
+            EXPECT_EQUAL(s2n_send(client_conn, app_data, sizeof(app_data), &blocked), sizeof(app_data));
+            size_t wiped_out_size = s2n_stuffer_data_available(&out);
+            EXPECT_EQUAL(original_out_size, wiped_out_size);
+        }
+
+        /* Handshake succeeds after wipe */
```

Comment:
Maybe we could have a test where the second handshake contains client auth? Since that's what renegotiate is for anyways.

Reply:
I wouldn't say that's "what renegotiate is for". It's a common use case, but not even the one we're adding this feature for. Client auth also doesn't depend on any of the variables I preserve through the wipe, so I'm not super worried about it. But test added anyway.